### PR TITLE
Issue 259: Document and revise parallel_execution_tbb 

### PR DESCRIPTION
**Pull request by jdgarciauc3m**

### CODE REVIEW

File: include/tbb/parallel_execution_tbb.h [hunk 1]
```diff
@@ -30,73 +30,124 @@
...
+  void set_queue_attributes(int size, queue_mode mode, int tokens) noexcept {
+    queue_size_ = size;
+    queue_mode_ = mode;
+    num_tokens_ = tokens;
+  }
+
+  /**
+  \brief Makes a communication queue for elements of type T.
+  Constructs a queue using the attributes that can be set via 
+  et_queue_attributes(). The value is returned via move semantics.
```
Comment (Drioastorga) [inline include/tbb/parallel_execution_tbb.h]: et_queue_attributes() -> set_queue_attributes()

File: include/native/parallel_execution_native.h [hunk 2]
```diff
@@ -32,105 +32,236 @@
...
-  queue_mode lockfree = queue_mode::blocking;
+/**
+\brief Thread index table to provide portable natural thread indices.
 
-  void set_queue_size(int new_size){
-     queue_size = new_size;
-  }
+A thread table provides a simple way to offer thread indices (starting from 0).
 
+When a thread registers itself in the registry. To get an integer index, users
```
Comment (Drioastorga) [inline include/native/parallel_execution_native.h]: When a thread registers itself in the registry. I think that this sentence should be: "When a thread is created, it regirsters itself in the registry" or "When a thread is created, it is registered by itself in the registry.".

File: include/native/parallel_execution_native.h [hunk 3]
```diff
@@ -32,105 +32,236 @@
...
-      thid_table.push_back(std::this_thread::get_id());    
-      lock.clear(std::memory_order_release);  
+\note This class is thread safe by means of using a spin-lock.
+*/
+class thread_registry {
+public:
+  thread_registry() noexcept = default;
+
+  /**
+  \brief Adds current thread in the registry.
```
Comment (Drioastorga) [inline include/native/parallel_execution_native.h]: Adds the current thread

File: include/native/parallel_execution_native.h [hunk 4]
```diff
@@ -32,105 +32,236 @@
...
+  */
+  void register_thread() noexcept;
+
+  /**
+  \brief Removes current thread from the registry.
```
Comment (Drioastorga) [inline include/native/parallel_execution_native.h]: Removes the current thread

File: include/native/parallel_execution_native.h [hunk 5]
```diff
@@ -32,105 +32,236 @@
...
+  */
+  void deregister_thread() noexcept;
+
+  /**
+  \brief Integer index for current thread
```
Comment (Drioastorga) [inline include/native/parallel_execution_native.h]: I think that "Returns the index of the current thread." is clearer.

File: include/native/parallel_execution_native.h [hunk 6]
```diff
@@ -32,105 +32,236 @@
...
+   degree.
+  */
+  parallel_execution_native() noexcept :
+      parallel_execution_native{
+          static_cast<int>(2 * std::thread::hardware_concurrency()), 
+          true}
+  {}
+
+  /** 
+  \brief Constructs a native parallele execution policy.
```
Comment (Drioastorga) [inline include/native/parallel_execution_native.h]: parallele -> parallel

File: include/tbb/parallel_execution_tbb.h [hunk 7]
```diff
@@ -30,73 +30,124 @@
...
 
+/**
+\brief Metafunction that determines if type E is supported in the current build.
+\tparam Execution policy type.
+*/
 template <typename E>
 constexpr bool is_supported();
 
+/**
+\brief Specialization stating that parallel_execution_native is supported.
```
Comment (Drioastorga) [inline include/tbb/parallel_execution_tbb.h]: parallel_execution_native -> parallel_execution_tbb

**Review by Drioastorga — CHANGES_REQUESTED** [review]
Changes requested are inlined

File: include/native/parallel_execution_native.h [hunk 8]
```diff
@@ -32,105 +32,236 @@
...
+  /**
+  \brief Get index of current thread in the thread table
+  \pre The current thread is currently registered.
+  */
+  int get_thread_id() const noexcept {
+    return thread_registry_.current_index();
+  }
+  
+  /**
+  \brief Sets the attributes for the queues built through make_queue<T<(()
```
Comment (mdolz) [inline include/native/parallel_execution_native.h]: Replace `make_queue<T<(()` by `make_queue<T>()`;

File: include/native/parallel_execution_native.h [hunk 9]
```diff
@@ -32,105 +32,236 @@
...
+  */
+  void set_queue_attributes(int size, queue_mode mode) noexcept {
+    queue_size_ = size;
+    queue_mode_ = mode;
+  }
+
+  /**
+  \brief Makes a communication queue for elements of type T.
+  Constructs a queue using the attributes that can be set via 
+  et_queue_attributes(). The value is returned via move semantics.
```
Comment (mdolz) [inline include/native/parallel_execution_native.h]: Replace `et` by `set`.

File: include/tbb/parallel_execution_tbb.h [hunk 10]
```diff
@@ -109,26 +160,37 @@ constexpr bool is_supported<parallel_execution_tbb>() {
...
+
 /// Parallel execution policy.
 /// Empty type if GRPPI_TBB disabled.
 struct parallel_execution_tbb {};
 
-/// Determine if a type is a TBB execution policy.
-/// False if GRPPI_TBB disabled.
+/**
+\brief Metafunction that determines if type E is parallel_execution_tbb
+This metafunction evaluates to fals if GRPPI_TBB is disabled.
```
Comment (javidoc00) [inline include/tbb/parallel_execution_tbb.h]: fals -> false

File: include/tbb/parallel_execution_tbb.h [hunk 11]
```diff
@@ -30,73 +30,124 @@
...
+/**
+\brief Metafunction that determines if type E is supported in the current build.
+\tparam Execution policy type.
+*/
 template <typename E>
 constexpr bool is_supported();
 
+/**
+\brief Specialization stating that parallel_execution_native is supported.
+This metafunction evaluates to fals if GRPPI_TBB is enabled.
```
Comment (javidoc00) [inline include/tbb/parallel_execution_tbb.h]: fals -> false

File: include/tbb/parallel_execution_tbb.h [hunk 12]
```diff
@@ -30,73 +30,124 @@
 
 namespace grppi{
 
-/** @brief Set the execution mode to parallel with threading building blocks
- *    (TBB) framework implementation
+/** 
+ \brief TBB parallel execution policy.
```
Comment (mdolz) [inline include/tbb/parallel_execution_tbb.h:34]: Shouldn't you say Intel TBB? This is just a minor issue, but the sooner it is solved the better.

File: include/tbb/parallel_execution_tbb.h [hunk 13]
```diff
@@ -30,73 +30,124 @@
...
+  void disable_ordering() noexcept { ordering_=false; }
 
   /**
   \brief Is execution ordered.
   */
-  bool is_ordered() const noexcept { return ordering; }
+  bool is_ordered() const noexcept { return ordering_; }
+
+  /**
+  \brief Sets the attributes for the queues built through make_queue<T<(()
```
Comment (mdolz) [inline include/tbb/parallel_execution_tbb.h]: Replace `make_queue<T<(()` by `make_queue<T>()`;

File: include/native/farm.h [hunk 14]
```diff
@@ -57,21 +57,19 @@ void farm(parallel_execution_native & ex, Generator generate_op,
...
   using namespace std;
   using result_type = typename result_of<Generator()>::type;
-  mpmc_queue<result_type> queue{ex.queue_size,ex.lockfree};
+  auto queue = ex.make_queue<result_type>();
 
   vector<thread> tasks;
   for (int i=0; i<ex.concurrency_degree(); ++i) {
     tasks.emplace_back([&](){
-      ex.register_thread();
+      auto manger = ex.thread_manager();
```
Comment (mdolz) [inline include/native/farm.h]: What does it mean `manger`? I'd rather prefer `manager`.

Comment (mdolz) [reply include/native/farm.h]: Indeed, I don't see where this variable is used inside the thread code...

File: include/native/parallel_execution_native.h [hunk 15]
```diff
@@ -32,105 +32,236 @@
...
+  while (lock_.test_and_set(memory_order_acquire)) {}
+  auto this_id = this_thread::get_id();
+  auto current = find(begin(ids_), end(ids_), this_id);
+  auto index = distance(begin(ids_), current);
+  lock_.clear(memory_order_release);
+  return index;
+};
+
+/**
+\brief RAII class to mange registration/deregistration pairs.
```
Comment (mdolz) [inline include/native/parallel_execution_native.h]: Replace `mange` by `manage`.

File: include/tbb/parallel_execution_tbb.h (hunk 10 above)
Comment (mdolz) [inline include/tbb/parallel_execution_tbb.h]: Replace `fals` by `false`.

**Review by javidoc00 — CHANGES_REQUESTED** [review]
See comments

**Review by mdolz — CHANGES_REQUESTED** [review]
See comments.

File: include/native/farm.h (hunk 14 above)
Comment (jdgarciauc3m) [reply include/native/farm.h]: Indeed, I change the name to manager.

It is not meant to be used. It is a RAII object. Consutructor acquires resource and destructor releases resource.

File: include/native/parallel_execution_native.h (hunk 2 above)
Comment (jdgarciauc3m) [reply include/native/parallel_execution_native.h]: Rewritten as:

A thread table provides a simple way to offer thread indices (starting from 0).

When a thread registers itself in the registry, its id is added to the vector 
of identifiers. When a thread deregisters itself from the registry its entry
is modified to contain the empty thread id.

To get an integer index, users may call `current_index`, which provides the order
number of the calling thread in the registry.

File: include/native/parallel_execution_native.h (hunk 3 above)
Comment (jdgarciauc3m) [reply include/native/parallel_execution_native.h]: Done

File: include/native/parallel_execution_native.h (hunk 4 above)
Comment (jdgarciauc3m) [reply include/native/parallel_execution_native.h]: Done

File: include/native/parallel_execution_native.h (hunk 5 above)
Comment (jdgarciauc3m) [reply include/native/parallel_execution_native.h]: See the `\returns` clause

File: include/native/parallel_execution_native.h (hunk 6 above)
Comment (jdgarciauc3m) [reply include/native/parallel_execution_native.h]: Done

File: include/native/parallel_execution_native.h (hunk 15 above)
Comment (jdgarciauc3m) [reply include/native/parallel_execution_native.h]: Done

File: include/native/parallel_execution_native.h (hunk 8 above)
Comment (jdgarciauc3m) [reply include/native/parallel_execution_native.h]: Done

File: include/native/parallel_execution_native.h (hunk 9 above)
Comment (jdgarciauc3m) [reply include/native/parallel_execution_native.h]: Done

File: include/tbb/parallel_execution_tbb.h (hunk 12 above)
Comment (jdgarciauc3m) [reply include/tbb/parallel_execution_tbb.h:34]: I use TBB in the short description and "Intel TBB" in the long description.

I think that should be enough.

File: include/tbb/parallel_execution_tbb.h (hunk 13 above)
Comment (jdgarciauc3m) [reply include/tbb/parallel_execution_tbb.h]: Done

File: include/tbb/parallel_execution_tbb.h (hunk 1 above)
Comment (jdgarciauc3m) [reply include/tbb/parallel_execution_tbb.h]: Done

File: include/tbb/parallel_execution_tbb.h (hunk 7 above)
Comment (jdgarciauc3m) [reply include/tbb/parallel_execution_tbb.h]: Done

File: include/tbb/parallel_execution_tbb.h (hunk 11 above)
Comment (jdgarciauc3m) [reply include/tbb/parallel_execution_tbb.h]: Done

File: include/tbb/parallel_execution_tbb.h (hunk 10 above)
Comment (jdgarciauc3m) [reply include/tbb/parallel_execution_tbb.h]: Done

Comment (jdgarciauc3m) [reply include/tbb/parallel_execution_tbb.h]: Done

**Comment by jdgarciauc3m** [comment]
Latest commit should resolve all code review comments.

**Review by javidoc00 — APPROVED** [review]

**Review by mdolz — APPROVED** [review]

**Review by Drioastorga — APPROVED** [review]

**Comment by jdgarciauc3m** [comment]
Closes issue #259

**Pull request merged**